### PR TITLE
642 move the removal of filter qa to imputation

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

Part 1: Remove _prev cols from full_estimation output

At the end of the imputation module, after outputting the imputation QA csv, we drop many of the imputation QA columns, including the  _ imputed cols, but we don't drop the _ prev columns and _ link columns used for MoR. 

This should be done so that the full_estimation dataframe isn't so large (and doesn't take so long to output).

Part 2: Move the removal of "no_imputation" and "no_mean_found" from outputs

This should now go at the end of imputation

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #642

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/imputation/imputation_helpers.py [hunk 1]
```diff
@@ -219,3 +219,54 @@ def fill_sf_zeros(df: pd.DataFrame) -> pd.DataFrame:
...
+
+    # Keep only clear and imputed records
+    imputed_statuses = ["TMI", "CF", "MoR", "constructed"]
+    to_keep = df["imp_marker"].isin(imputed_statuses) | (df["imp_marker"] == "R")
+
+    to_keep_df = df.copy().loc[to_keep]
+    filtered_output_df = df.copy().loc[~to_keep]
+
+    # change the value of the status column to 'imputed' for imputed statuses
+    condition = to_keep_df["status"].isin(imputed_statuses)
```
Comment (AnneONS) [inline src/imputation/imputation_helpers.py]: should be imp_marker

File: src/imputation/imputation_helpers.py [hunk 2]
```diff
@@ -219,3 +219,54 @@ def fill_sf_zeros(df: pd.DataFrame) -> pd.DataFrame:
...
+        for col in df.columns
+        if (col.endswith("prev") | col.endswith("imputed") | col.endswith("link"))
+    ]
+
+    to_drop += ["200_original", "pg_sic_class", "empty_pgsic_group", "empty_pg_group"]
+    to_drop += ["200_imp_marker", "211_trim", "305_trim", "manual_trim"]
+    df = df.drop(columns=to_drop)
+
+    # Keep only clear and imputed records
+    imputed_statuses = ["TMI", "CF", "MoR", "constructed"]
```
Comment (AnneONS) [inline src/imputation/imputation_helpers.py]: imp_markers_to_keep

File: src/imputation/imputation_helpers.py [hunk 3]
```diff
@@ -219,3 +219,54 @@ def fill_sf_zeros(df: pd.DataFrame) -> pd.DataFrame:
         df.loc[(sf_mask & clear_mask), q] = df.copy()[q].fillna(0)
 
     return df
+
+
+def tidy_imputation_dataframe(
+        df: pd.DataFrame,
+        config: Dict,
+        logger,
```
Comment (AnneONS) [inline src/imputation/imputation_helpers.py]: what object?

**Review by AnneONS — COMMENTED** [review]
my own comments!

File: src/imputation/imputation_helpers.py (hunk 3 above)
Comment (jwestw) [reply src/imputation/imputation_helpers.py]: `logging.Logger` I think

File: src/imputation/imputation_helpers.py (hunk 2 above)
Comment (jwestw) [reply src/imputation/imputation_helpers.py]: Changed

File: src/imputation/imputation_helpers.py (hunk 3 above)
Comment (jwestw) [reply src/imputation/imputation_helpers.py]: changed

File: src/outputs/form_output_prep.py [hunk 4]
```diff
@@ -28,26 +28,13 @@ def form_output_prep(
...
-    # filter estimated_df and weighted_df to only include clear or imputed statuses
-    outputs_df = estimated_df.copy().loc[to_keep]
-    tau_outputs_df = weighted_df.copy().loc[to_keep]
-
-    # filter estimated_df for records not included in outputs
-    filtered_output_df = estimated_df.copy().loc[~to_keep]
+    outputs_df = estimated_df.copy().loc[no_rnd_spenders_filter]
+    tau_outputs_df = weighted_df.copy().loc[no_rnd_spenders_filter]
 
     if ni_full_responses is not None:
```
Comment (jwestw) [inline src/outputs/form_output_prep.py:39]: I know this is not part of you ticket Anne, but I would recommend moving this NI processing to another function.

File: src/outputs/form_output_prep.py [hunk 5]
```diff
@@ -66,19 +53,10 @@ def form_output_prep(
         # outputs_df = pd.concat([outputs_df, ni_full_responses])
         tau_outputs_df = pd.concat([tau_outputs_df, ni_full_responses])
```
Comment (jwestw) [inline src/outputs/form_output_prep.py:54]: This line removes the need for a logic gate before hand. 

`tau_outputs_df = pd.concat([tau_outputs_df, ni_full_responses]) if ni_full_responses is not None else tau_outputs_df`

Then we won't need the conditional block and makes the code more readable and computationally efficient.

Again, I know this was not part of your ticket but we may as well generate nice to have code-tidying and efficiency tickets while I am looking at the code.

File: src/outlier_detection/auto_outliers.py [hunk 6]
```diff
@@ -238,14 +238,6 @@ def run_auto_flagging(
 
     # loop through all columns to be flagged for outliers
     for value_col in flag_value_cols:
-        # to_numeric is needed to convert strings. However 'coerce'
-        # means values that
-        # can't be converted are represented by NaNs.
-        # TODO data validation and cleaning should replace the need for
-        # 'to_numeric'
-        # check ticket (RDRP-386)
-        df[value_col] = pd.to_numeric(df[value_col], errors="coerce")
```
Comment (jwestw) [inline src/outlier_detection/auto_outliers.py:247]: Great that this TODO has been resolved.

**Review by jwestw — APPROVED** [review]
This all looks good and the pipeline runs fine.

**Pull request merged**